{% extends "base2.html" %}
{% block title %}Network Portal - About{% endblock %}
{% block content %}

<div class="sub-page">
  <h5>About the Network Portal</h5>
  <p>version: {{version}}</p>
  <p>The Network Portal is a database of gene transcription regulatory networks, 
    enabling exploration, annotation and comparative analysis. Deciphering the complexity of biological systems requires system level view of the regulatory players and their interactions. However, complexity of these interactions and uncertainity in their inference present a challenge for analysis and biological discovery. Furthermore, visualizing and analyzing these interactions often requires advanced computational expertise, tools and resources. Network Portal provides analysis and visualization tools for selected gene regulatory networks to aid researchers in biological discovery and hypothesis development.</p>
  <h5>Method</h5>
  <p> Network Portal integrates identification of conditionally co-regulated group of genes by using <a href="http://baliga.systemsbiology.net/cmonkey">cMonkey</a> and inference of regulatory influences by using <a href="http://baliga.systemsbiology.net/drupal/content/inferelator">Inferelator</a>. In addition, it is integrated with <a href="http://microbesonline.org">MicrobesOnline</a> for genomic information and <a href="http://regprecise.lbl.gov/RegPrecise">RegPrecise</a> for collection of manually curatedregulatory inferences.</p>
The Network Portal aims to be flexible, modular and comprehensive. Therefore we designed our framework in two interconnected layers.<br>
<p><strong>1. Automated Network Inference Pipeline:</strong> The first phase of this module leverages various data (microarray, RNAseq etc.) from validated resources (GEO, MicrobesOnline, ArrayExpress etc), checks for quality control and perform necessary optimizations to ensure compliance with <strong>ENIGMA Standards and Best Practises</strong>. The second phase of the module integrates algorithms to infer regulatory networks. At this stage two algorithms, cMonkey and Inferelator are integrated with the portal. Upcoming releases will allow developers and researchers to plugin their favorite inference algorithm seamlessly without worrying about data formats or downstream applications. In the final phase of the automated inference pipeline, processed output is transmitted to database that can be accessed pprogrammatically within and outside of the portal.</p>  
<p><strong>2. Visualisation and Analysis </strong> This layer provides online tools for visualising network information in the context of other available data (such as genomics, proteomics, orthology etc.) in order to help with hypothesis validation and creation. Similar to automated Inference Pipeline, this layer is also designed to be modular and flexible. We already provide tools that highlight important connections between various types of data and network information. Developers are also encouraged to design visualisation and data harvest tools that can be plugged in to our visualisation framework. For example, in addition to showing regulatory influences for a particular gene, a <strong><a href="http://regprecise.lbl.gov/RegPrecise/">RegPrecise</a></strong> plugin can leverage verified regulatory informations and integrate with network portal data. Moreover, in order to give researchers more freedom, Network Portal is integrated with <a href="http://gaggle.systemsbiology.net">GAGGLE</a> Framework which enables broadcast of data among various desktop and online tools.  
  <h5>Credits</h5>
  <p> The Network Portal was developed as part of the <a href="http://genomicscience.energy.gov/compbio/#page=news">DOE Knowledgebase</a> effort.  While the Network Portal integrates data, algorithms, software, and technologies that have been developed by the community, development of the portal itself is being driven by efforts focusing on the biology of specific organisms including:
    <ul>
      <li>Desulfovibrio vulgaris (funded by <a href="http://enigma.lbl.gov/">ENIGMA</a>) </li>
      <li>Methanococcus maripaludis (funded by <a href="http://enigma.lbl.gov/">ENIGMA</a> and DOE) </li> 
      <li>Halobacterium salinarum (funded by NSF, NIH and DOE) </li>
    </ul>

  <h5>Tools</h5>
  <p>Python, Django, NetworkX, Solr, PostgeSQL, Cytoscape Web, jQuery, the R project</p>
  <p>Data sources include: Microbes Online, KEGG, NCBI, GO, TIGR</p>
  
  <h5>Code</h5>
  <p>The <a href="http://baliga-lab/github.com/network_portal">source code for the Network Portal</a> is available on GitHub.</p>
  
  <h5>License</h5>
  <p>This software is released under the terms of the <a href="http://www.gnu.org/licenses/lgpl.html">LGPL</a>. Copyright (C) 2011 Institute for Systems Biology</p>
</div><!-- End #main -->
{% endblock %}
